---
title: "Transcriptome Demo"
author: "Tricia"
date: "`r Sys.Date()`"
output: pdf_document
---

# Load required packages (you might have to figure out how to install some of these first...)
```{r message=FALSE}
library(ballgown)
library(RColorBrewer)
library(genefilter)
library(dplyr)
library(devtools)
```

# Install Ballgown
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ballgown")
```

# This code is creating a table that sorts the phenotype of our samples by their growth stage
```{r}
pheno_data<-data.frame(ids = c("plank01", "plank02", "biofilm01", "biofilm02"),
                       stage = c("planktonic", "planktonic", "biofilm", "biofilm"))
```

# Create Ballgown object and check transcript number
```{r message=FALSE}
samples.c <- paste('ballgown', pheno_data$ids, sep = '/')
bg <- ballgown(samples = samples.c, meas='all', pData = pheno_data)
bg
```

# This code is filtering the transcripts of the ballgown folder made in the previous step in order to remove the reads that have low varience of expression (less than 1) across the samples. This allows us to only focus on the reads that have a meaningful amount of expression over the four samples.
```{r}
bg_filt = subset(bg,"rowVars(texpr(bg)) >1",genomesubset=TRUE)
bg_filt
```

# Create a table of transcripts
```{r}
results_transcripts<- stattest(bg_filt, feature = "transcript", covariate = "stage",
getFC = TRUE, meas = "FPKM")
results_transcripts<-data.frame(geneNames=geneNames(bg_filt),
transcriptNames=transcriptNames(bg_filt), results_transcripts)
```

# Choose a transcript to examine more closely
```{r}
results_transcripts[results_transcripts$transcriptNames == "gene-PA0004", ]
```
# The information I am given about this transcript is that it represents the gene gyrB and it is the fourth gene in the data set. The high fc value tells me that this gene is highly upregulated, but the p-value and q-value tell me that there is not a significant difference in the expression of this gene between the four samples.

# This code is filtering out any transcripts that do not have a significant difference (p-value less than 0.05) between the four samples. In other words we are keeping anything that has a statically signifanct difference between the samples.
```{r}
sigdiff <- results_transcripts %>% filter(pval<0.05)
dim(sigdiff)
```

# Organize the table. The table is being organized by the p-value from the lowest p-value to the highest.
```{r}
o = order(sigdiff[,"pval"], -abs(sigdiff[,"fc"]), decreasing=FALSE)
output = sigdiff[o,c("geneNames","transcriptNames", "id","fc","pval","qval")]
write.table(output, file="SigDiff.txt", sep="\t", row.names=FALSE, quote=FALSE)
head(output)
```

# Load gene names
```{r}
bg_table = texpr(bg_filt, 'all')
bg_gene_names = unique(bg_table[, 9:10])
```

# Pull out gene expression data and visualize
```{r}
gene_expression = as.data.frame(gexpr(bg_filt))
head(gene_expression)
```

# This code lets us view the first six rows of our gene_expression table and tells us the dimentions of the gene_expression table. (4609 rows, 4 columns/samples)
```{r}
colnames(gene_expression) <- c("plank01", "plank02", "biofilm01", "biofilm02")
head(gene_expression)
dim(gene_expression)
```

# Load the transcript to gene table and determine the number of transcripts and unique genes
```{r}
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)
length(row.names(transcript_gene_table))
length(unique(transcript_gene_table[,"g_id"]))
```

# Plot the number of transcripts per gene
```{r}
counts=table(transcript_gene_table[,"g_id"])
c_one = length(which(counts == 1))
c_more_than_one = length(which(counts > 1))
c_max = max(counts)
hist(counts, breaks=50, col="bisque4", xlab="Transcripts per gene",
main="Distribution of transcript count per gene")
legend_text = c(paste("Genes with one transcript =", c_one),
paste("Genes with more than one transcript =", c_more_than_one),
paste("Max transcripts for single gene = ", c_max))
legend("topright", legend_text, lty=NULL)
```

# Due to alternative splicing, there can be more than one transcript produced per gene. This table tells us that the majority of genes had less than five transcripts, but the number of transcripts for a gene could be as high as 64.

# Create a plot of how similar the two replicates are for one another. We have two data sets...how can you modify this code in another chunk to create a plot of the other set?(By replacing "plankonic" in the first code with "biofilm")
```{r}
x = gene_expression[,"plank01"]
y = gene_expression[,"plank02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="purple", cex=0.25,
xlab="FPKM (Planktonic, Replicate 1)", ylab="FPKM (Planktonic, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```
```{r}
x = gene_expression[,"biofilm01"]
y = gene_expression[,"biofilm02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="purple", cex=0.25,
xlab="FPKM (Biofilm, Replicate 1)", ylab="FPKM (Biofilm, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```

# If the two data sets are similar, then they have a similar level of gene expression as determined using the FPKM (expression level).

# Create plot of differential gene expression between the conditions
```{r}
results_genes = stattest(bg_filt, feature="gene", covariate="stage", getFC=TRUE, meas="FPKM")
results_genes = merge(results_genes,bg_gene_names,by.x=c("id"),by.y=c("gene_id"))
sig=which(results_genes$pval<0.05)
results_genes[,"de"] = log2(results_genes[,"fc"])
hist(results_genes[sig,"de"], breaks=50, col="seagreen",
xlab="log2(Fold change) Planktonic vs Biofilm",
main="Distribution of differential expression values")
abline(v=-2, col="black", lwd=2, lty=2)
abline(v=2, col="black", lwd=2, lty=2)
legend("topleft", "Fold-change > 4", lwd=2, lty=2)
```

# The figure generated from the code above show us the amount of genes upregulated in planktonic cells versus the biofilm cells. In this figure, there are more genes upregulated in the planktonic cells than the biofilm cells, but there are still a good amount of genes upregulated in biofilm cells. Most of the genes seem to be upregulated 20 fold in planktonic cells or biofilm cells.

# Plot total gene expression highlighting differentially expressed genes
```{r}
gene_expression[,"plank"]=apply(gene_expression[,c(1:2)], 1, mean)
gene_expression[,"biofilm"]=apply(gene_expression[,c(3:4)], 1, mean)
x=log2(gene_expression[,"plank"]+min_nonzero)
y=log2(gene_expression[,"biofilm"]+min_nonzero)
plot(x=x, y=y, pch=16, cex=0.25, xlab="Planktonic FPKM (log2)", ylab="Biofilm FPKM (log2)",
main="Planktonic vs Biofilm FPKMs")
abline(a=0, b=1)
xsig=x[sig]
ysig=y[sig]
points(x=xsig, y=ysig, col="magenta", pch=16, cex=0.5)
legend("topleft", "Significant", col="magenta", pch=16)
```

# Make a table of FPKM values
```{r}
fpkm = texpr(bg_filt,meas="FPKM")
```

# Choose a gene to determine individual expression
```{r}
ballgown::transcriptNames(bg_filt)[8]
ballgown::geneNames(bg_filt)[8]
```

# Transform to log2
```{r}
transformed_fpkm <- log2(fpkm[8, ] + 1)
```

# Make sure values are properly coded as numbers
```{r}
numeric_stages <- as.numeric(factor(pheno_data$stage))

jittered_stages <- jitter(numeric_stages)
```

# Plot expression of individual gene
```{r}
boxplot(transformed_fpkm ~ pheno_data$stage,
        main=paste(ballgown::geneNames(bg_filt)[8], ' : ', ballgown::transcriptNames(bg_filt)[8]),
        xlab="Stage",
        ylab="log2(FPKM+1)",
        col=c("lightblue", "salmon"),
        border="darkblue")

points(transformed_fpkm ~ jittered_stages, 
       pch=21, col="blue", bg="lightblue", cex=1.2)
```
# The figure generated from this code shows that the expression of glyS is more upregulated (higher expresssion) in the planktonic sample than the biofilm sample.